skeleton/src/main/java/com/dropbox/kaiken/skeleton/skeleton/core/KaikenUserServicesProvider.kt

```diff
@@ -0,0 +1,67 @@
+package com.dropbox.kaiken.skeleton.skeleton.core
+
+import com.dropbox.kaiken.scoping.AppServices
+import com.dropbox.kaiken.scoping.TeardownHelper
+import com.dropbox.kaiken.scoping.UserServices
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+/**
+ * Skeleton implementation of [UserServicesProvider] that registers itself with a [UserManager].
+ *
+ */
+class KaikenUserServicesProvider
+constructor(
+        private val applicationServices: AppServices,
+        private val userServicesFactory: (AppServices, SkeletonUser) -> UserServices,
+) : SkeletonUserServicesProvider {
+    private val userServicesMap = mutableMapOf<String, KaikenUserServices>()
+
+    init {
+        //TODO Mike figure out if we want another scope to launch from
+        GlobalScope.launch {
+            (applicationServices as KaikenAppServices).userManager().getUserState().collect { userState ->
+                userState.usersRemoved.forEach { user ->
+                    teardownUserServicesOf(user.userId)
+                }
+                //TODO Mike figure out what we want to do for the base user type
+                userState.usersAdded.forEach { user ->
+                    initUserServicesOf(
+                            SkeletonUser(
+                                    user.userId,
+                                    SkeletonOauth2(user.accessToken),
+                                    user.isActiveUser
+                            )
+                    )
+                }
+            }
+        }
+    }
+
+    override fun initUserServicesOf(user: SkeletonUser) {
+        synchronized(userServicesMap) {
+            val userServices: KaikenUserServices = userServicesFactory(applicationServices, user) as KaikenUserServices
+            userServicesMap[user.userId] = userServices
+        }
+    }
+
+    override fun teardownUserServicesOf(userId: String) {
+        synchronized(userServicesMap) {
+            val teardownHelper = userServicesMap.remove(userId)?.getTeardownHelper()
+            if (teardownHelper != null) {
+                dispatchTeardown(teardownHelper)
```

Comment:
Ah! This solves my previous Teardown question.

Reply:
I actually needed to split the teardown helpers into separate types. This was the negative consequence of subcomponents (all types from parents get provided into child)